File: app/src/main/java/com/o4x/musical/ui/fragments/settings/SettingsFragment.java
package com.o4x.musical.ui.fragments.settings;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.media.audiofx.AudioEffect;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceManager;
import androidx.preference.TwoStatePreference;

import com.afollestad.materialdialogs.color.ColorChooserDialog;
import com.kabouzeid.appthemehelper.ThemeStore;
import com.kabouzeid.appthemehelper.common.prefs.supportv7.ATEColorPreference;
import com.kabouzeid.appthemehelper.common.prefs.supportv7.ATEPreferenceFragmentCompat;
import com.kabouzeid.appthemehelper.util.ColorUtil;
import com.o4x.musical.App;
import com.o4x.musical.R;
import com.o4x.musical.appshortcuts.DynamicShortcutManager;
import com.o4x.musical.preferences.BlacklistPreference;
import com.o4x.musical.preferences.BlacklistPreferenceDialog;
import com.o4x.musical.preferences.LibraryPreference;
import com.o4x.musical.preferences.LibraryPreferenceDialog;
import com.o4x.musical.preferences.NowPlayingScreenPreference;
import com.o4x.musical.preferences.NowPlayingScreenPreferenceDialog;
import com.o4x.musical.ui.activities.PurchaseActivity;
import com.o4x.musical.util.NavigationUtil;
import com.o4x.musical.util.PreferenceUtil;

public class SettingsFragment extends ATEPreferenceFragmentCompat implements SharedPreferences.OnSharedPreferenceChangeListener {

    private static void setSummary(@NonNull Preference preference) {
        setSummary(preference, PreferenceManager
                .getDefaultSharedPreferences(preference.getContext())
                .getString(preference.getKey(), ""));
    }

    private static void setSummary(Preference preference, @NonNull Object value) {
        String stringValue = value.toString();

        if (preference instanceof ListPreference) {
            ListPreference listPreference = (ListPreference) preference;
            int index = listPreference.findIndexOfValue(stringValue);
            preference.setSummary(
                    index >= 0
                            ? listPreference.getEntries()[index]
                            : null);
        } else {
            preference.setSummary(stringValue);
        }
    }

    @Override
    public void onCreatePreferences(Bundle bundle, String s) {
        addPreferencesFromResource(R.xml.pref_library);
        addPreferencesFromResource(R.xml.pref_colors);
        addPreferencesFromResource(R.xml.pref_notification);
        addPreferencesFromResource(R.xml.pref_now_playing_screen);
        addPreferencesFromResource(R.xml.pref_images);
        addPreferencesFromResource(R.xml.pref_lockscreen);
        addPreferencesFromResource(R.xml.pref_audio);
        addPreferencesFromResource(R.xml.pref_playlists);
        addPreferencesFromResource(R.xml.pref_blacklist);
    }

    @Nullable
    @Override
    public DialogFragment onCreatePreferenceDialog(Preference preference) {
        if (preference instanceof NowPlayingScreenPreference) {
            return NowPlayingScreenPreferenceDialog.newInstance();
        } else if (preference instanceof BlacklistPreference) {
            return BlacklistPreferenceDialog.newInstance();
        } else if (preference instanceof LibraryPreference) {
            return LibraryPreferenceDialog.newInstance();
        }
        return super.onCreatePreferenceDialog(preference);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getListView().setPadding(0, 0, 0, 0);
        invalidateSettings();
        PreferenceUtil.getInstance(getActivity()).registerOnSharedPreferenceChangedListener(this);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        PreferenceUtil.getInstance(getActivity()).unregisterOnSharedPreferenceChangedListener(this);
    }

    public void invalidateSettings() {
        final Preference generalTheme = findPreference("general_theme");
        setSummary(generalTheme);
        generalTheme.setOnPreferenceChangeListener((preference, o) -> {
            String themeName = (String) o;
            if (themeName.equals("black") && !App.isProVersion()) {
                Toast.makeText(getActivity(), R.string.black_theme_is_a_pro_feature, Toast.LENGTH_LONG).show();
                startActivity(new Intent(getContext(), PurchaseActivity.class));
                return false;
            }

            setSummary(generalTheme, o);

            ThemeStore.markChanged(getActivity());

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
                // Set the new theme so that updateAppShortcuts can pull it
                getActivity().setTheme(PreferenceUtil.getThemeResFromPrefValue(themeName));
                new DynamicShortcutManager(getActivity()).updateDynamicShortcuts();
            }

            getActivity().recreate();
            return true;
        });

        final Preference autoDownloadImagesPolicy = findPreference("auto_download_images_policy");
        setSummary(autoDownloadImagesPolicy);
        autoDownloadImagesPolicy.setOnPreferenceChangeListener((preference, o) -> {
            setSummary(autoDownloadImagesPolicy, o);
            return true;
        });

        final ATEColorPreference primaryColorPref = (ATEColorPreference) findPreference("primary_color");
        final int primaryColor = ThemeStore.primaryColor(getActivity());
        primaryColorPref.setColor(primaryColor, ColorUtil.darkenColor(primaryColor));
        primaryColorPref.setOnPreferenceClickListener(preference -> {
            new ColorChooserDialog.Builder(getActivity(), R.string.primary_color)
                    .accentMode(false)
                    .allowUserColorInput(true)
                    .allowUserColorInputAlpha(false)
                    .preselect(primaryColor)
                    .show(getActivity());
            return true;
        });

        final ATEColorPreference accentColorPref = (ATEColorPreference) findPreference("accent_color");
        final int accentColor = ThemeStore.accentColor(getActivity());
        accentColorPref.setColor(accentColor, ColorUtil.darkenColor(accentColor));
        accentColorPref.setOnPreferenceClickListener(preference -> {
            new ColorChooserDialog.Builder(getActivity(), R.string.accent_color)
                    .accentMode(true)
                    .allowUserColorInput(true)
                    .allowUserColorInputAlpha(false)
                    .preselect(accentColor)
                    .show(getActivity());
            return true;
        });

        TwoStatePreference colorNavBar = (TwoStatePreference) findPreference("should_color_navigation_bar");
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            colorNavBar.setVisible(false);
        } else {
            colorNavBar.setChecked(ThemeStore.coloredNavigationBar(getActivity()));
            colorNavBar.setOnPreferenceChangeListener((preference, newValue) -> {
                ThemeStore.editTheme(getActivity())
                        .coloredNavigationBar((Boolean) newValue)
                        .commit();
                getActivity().recreate();
                return true;
            });
        }

        final TwoStatePreference classicNotification = (TwoStatePreference) findPreference("classic_notification");
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            classicNotification.setVisible(false);
        } else {
            classicNotification.setChecked(PreferenceUtil.getInstance(getActivity()).classicNotification());
            classicNotification.setOnPreferenceChangeListener((preference, newValue) -> {
                // Save preference
                PreferenceUtil.getInstance(getActivity()).setClassicNotification((Boolean) newValue);
                return true;
            });
        }

        final TwoStatePreference coloredNotification = (TwoStatePreference) findPreference("colored_notification");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            coloredNotification.setEnabled(PreferenceUtil.getInstance(getActivity()).classicNotification());
        } else {
            coloredNotification.setChecked(PreferenceUtil.getInstance(getActivity()).coloredNotification());
            coloredNotification.setOnPreferenceChangeListener((preference, newValue) -> {
                // Save preference
                PreferenceUtil.getInstance(getActivity()).setColoredNotification((Boolean) newValue);
                return true;
            });
        }

        final TwoStatePreference colorAppShortcuts = (TwoStatePreference) findPreference("should_color_app_shortcuts");
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N_MR1) {
            colorAppShortcuts.setVisible(false);
        } else {
            colorAppShortcuts.setChecked(PreferenceUtil.getInstance(getActivity()).coloredAppShortcuts());
            colorAppShortcuts.setOnPreferenceChangeListener((preference, newValue) -> {
                // Save preference
                PreferenceUtil.getInstance(getActivity()).setColoredAppShortcuts((Boolean) newValue);

                // Update app shortcuts
                new DynamicShortcutManager(getActivity()).updateDynamicShortcuts();

                return true;
            });
        }

        final Preference equalizer = findPreference("equalizer");
        if (!hasEqualizer()) {
            equalizer.setEnabled(false);
            equalizer.setSummary(getResources().getString(R.string.no_equalizer));
        }
        equalizer.setOnPreferenceClickListener(preference -> {
            NavigationUtil.openEqualizer(getActivity());
            return true;
        });

        updateNowPlayingScreenSummary();
    }

    private boolean hasEqualizer() {
        final Intent effects = new Intent(AudioEffect.ACTION_DISPLAY_AUDIO_EFFECT_CONTROL_PANEL);
        PackageManager pm = getActivity().getPackageManager();
        ResolveInfo ri = pm.resolveActivity(effects, 0);
        return ri != null;
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        switch (key) {
            case PreferenceUtil.NOW_PLAYING_SCREEN_ID:
                updateNowPlayingScreenSummary();
                break;
            case PreferenceUtil.CLASSIC_NOTIFICATION:
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    findPreference("colored_notification").setEnabled(sharedPreferences.getBoolean(key, false));
                }
                break;
        }
    }

    private void updateNowPlayingScreenSummary() {
        findPreference("now_playing_screen_id").setSummary(PreferenceUtil.getInstance(getActivity()).getNowPlayingScreen().titleRes);
    }
}

File: app/src/main/java/com/o4x/musical/ui/fragments/settings/SettingsFragment.kt
package com.o4x.musical.ui.fragments.settings

import android.content.Intent
import android.content.SharedPreferences
import android.media.audiofx.AudioEffect
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.navigation.fragment.findNavController
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.TwoStatePreference
import com.o4x.appthemehelper.ThemeStore
import com.o4x.appthemehelper.extensions.backgroundColor
import com.o4x.musical.R
import com.o4x.musical.helper.MusicPlayerRemote
import com.o4x.musical.preferences.LibraryPreferenceDialog
import com.o4x.musical.ui.dialogs.DeleteCachedDialog
import com.o4x.musical.ui.dialogs.DeleteCustomImagesDialog
import com.o4x.musical.ui.dialogs.SmartPlaylistLimitDialog
import com.o4x.musical.util.NavigationUtil
import com.o4x.musical.prefs.PreferenceUtil

class SettingsFragment : PreferenceFragmentCompat(), SharedPreferences.OnSharedPreferenceChangeListener {

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        PreferenceUtil.registerOnSharedPreferenceChangedListener(this)

        listView.setBackgroundColor(backgroundColor())
        invalidateSettings()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        PreferenceUtil.unregisterOnSharedPreferenceChangedListener(this)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.prefs)
    }

    private fun invalidateSettings() {

          ////////////////////
         // THEME SETTINGS //
        ////////////////////
        val generalTheme: Preference? = findPreference(PreferenceUtil.GENERAL_THEME)
        var lastTheme = generalTheme?.summary.toString()
        generalTheme?.setOnPreferenceChangeListener { _, newValue ->
            val newTheme = newValue.toString()
            if (lastTheme != newTheme) {
                lastTheme = newTheme
                ThemeStore.markChanged(requireContext())
            }
            true
        }

        val themeColorPref: Preference = findPreference(PreferenceUtil.THEME_COLOR)!!
        themeColorPref.isVisible = PreferenceUtil.isDarkMode
        var lastThemeColor = themeColorPref.summary.toString()
        themeColorPref.setOnPreferenceChangeListener { _, newValue ->
            val newThemeColor = newValue.toString()
            if (lastThemeColor != newThemeColor) {
                lastThemeColor = newThemeColor
                ThemeStore.markChanged(requireContext())
            }
            true
        }

          ////////////////////////
         // INTERFACE SETTINGS //
        ////////////////////////

        val libraryPref: Preference = findPreference(PreferenceUtil.LIBRARY_CATEGORIES)!!
        libraryPref.setOnPreferenceClickListener {
            val fragment = LibraryPreferenceDialog.newInstance()
            fragment.show(childFragmentManager, libraryPref.key)
            return@setOnPreferenceClickListener true
        }

          ////////////////////
         // IMAGE SETTINGS //
        ////////////////////

        val deleteCachedPref: Preference = findPreference(PreferenceUtil.DELETE_CACHED_IMAGES)!!
        deleteCachedPref.setOnPreferenceClickListener {
            DeleteCachedDialog.create().show(
                childFragmentManager, it.key
            )
            return@setOnPreferenceClickListener true
        }

        val deleteCustomPref: Preference = findPreference(PreferenceUtil.DELETE_CUSTOM_IMAGES)!!
        deleteCustomPref.setOnPreferenceClickListener {
            DeleteCustomImagesDialog.create().show(
                childFragmentManager, it.key
            )
            return@setOnPreferenceClickListener true
        }

          ////////////////////
         // AUDIO SETTINGS //
        ////////////////////

        val findPreference: Preference = findPreference(PreferenceUtil.EQUALIZER)!!
        if (!hasEqualizer()) {
            findPreference.isEnabled = false
            findPreference.summary = resources.getString(R.string.no_equalizer)
        } else {
            findPreference.isEnabled = true
        }
        findPreference.setOnPreferenceClickListener {
            NavigationUtil.openEqualizer(requireActivity())
            true
        }

          ///////////////////////////
         // NOTIFICATION SETTINGS //
        ///////////////////////////

        val classicNotification: TwoStatePreference? = findPreference(PreferenceUtil.CLASSIC_NOTIFICATION)
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            classicNotification?.isVisible = false
        } else {
            classicNotification?.apply {
                isChecked = PreferenceUtil.isClassicNotification
                setOnPreferenceChangeListener { _, newValue ->
                    // Save preference
                    PreferenceUtil.isClassicNotification = newValue as Boolean
                    invalidateSettings()
                    true
                }
            }
        }

        val coloredNotification: TwoStatePreference? = findPreference(PreferenceUtil.COLORED_NOTIFICATION)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            coloredNotification?.isEnabled = PreferenceUtil.isClassicNotification
        } else {
            coloredNotification?.apply {
                isChecked = PreferenceUtil.isColoredNotification
                setOnPreferenceChangeListener { _, newValue ->
                    PreferenceUtil.isColoredNotification = newValue as Boolean
                    true
                }
            }
        }

          ///////////////////////
         // ADVANCED SETTINGS //
        ///////////////////////

        val smartPlaylistPreference: Preference = findPreference(PreferenceUtil.SMART_PLAYLIST_LIMIT)!!
        setSummary(smartPlaylistPreference, PreferenceUtil.smartPlaylistLimit)
        smartPlaylistPreference.setOnPreferenceClickListener {
            SmartPlaylistLimitDialog.create().show(
                childFragmentManager, it.key
            )
            return@setOnPreferenceClickListener true
        }

        val languagePreference: Preference? = findPreference(PreferenceUtil.LANGUAGE_NAME)
        languagePreference?.setOnPreferenceChangeListener { prefs, newValue ->
            setSummary(prefs, newValue)
            true
        }

        val aboutPreference: Preference? = findPreference(PreferenceUtil.ABOUT)
        aboutPreference?.setOnPreferenceClickListener {
            try {
                findNavController().navigate(R.id.action_mainSettings_to_about)
            } catch (e: Exception) {}
            true
        }
    }


    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        when (key) {
            PreferenceUtil.CLASSIC_NOTIFICATION -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    findPreference<Preference>(PreferenceUtil.COLORED_NOTIFICATION)?.isEnabled =
                        sharedPreferences.getBoolean(key, false)
                }
            }
            PreferenceUtil.SMART_PLAYLIST_LIMIT -> {
                findPreference<Preference>(PreferenceUtil.SMART_PLAYLIST_LIMIT)?.summary =
                    PreferenceUtil.smartPlaylistLimit.toString()
            }
            PreferenceUtil.FILTER_SONG,
            PreferenceUtil.COLORED_FOOTER,
            PreferenceUtil.IGNORE_MEDIA -> MusicPlayerRemote.notifyMediaStoreChanged()
        }
    }


    private fun hasEqualizer(): Boolean {
        val effects = Intent(AudioEffect.ACTION_DISPLAY_AUDIO_EFFECT_CONTROL_PANEL)

        val pm = requireActivity().packageManager
        val ri = pm.resolveActivity(effects, 0)
        return ri != null
    }

    private fun setSummary(preference: Preference, value: Any?) {
        val stringValue = value.toString()
        if (preference is ListPreference) {
            val index = preference.findIndexOfValue(stringValue)
            preference.setSummary(if (index >= 0) preference.entries[index] else null)
        } else {
            preference.summary = stringValue
        }
    }
}